src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,225 @@
+        // for now, get rid of this eventually
+        sign_(std::move(sign)) {}
+
+  const std::string& name() const { return name_; }
+  const std::vector<std::string>& args() const { return args_; }
+  Sign sign() { return sign_; }
+
+ private:
+  std::string name_;
+  std::vector<std::string> args_;
+  Sign sign_;
+};
+
+// Attribute corresponds to an attribute of a principal with the form
+// <Principal> <pred_name>(arg_1, ..., arg_n)
+class Attribute {
+ public:
+  explicit Attribute(Principal principal, Predicate predicate)
+    : principal_(principal), predicate_(predicate) {}
+  const Principal& principal() const { return principal_; }
+  const Predicate& predicate() const { return predicate_; }
+ private:
+  Principal principal_;
+  Predicate predicate_;
+};
+
+// CanActAs corresponds to a canActAs expression of the form
+// <principal_1> canActAs <principal_2>
+class CanActAs {
+ public:
+  explicit CanActAs(Principal left_principal, Principal right_principal)
+    : left_principal_(left_principal),
+      right_principal_(right_principal) {}
+  const Principal& left_principal() const { return left_principal_; }
+  const Principal& right_principal() const { return right_principal_; }
+
+ private:
+   Principal left_principal_;
+   Principal right_principal_;
+};
+
+// BaseFact corresponds to a base-case fact that cannot recursively include other 
+// facts through "canSay". "Fact" includes the recursive case. It is necessary 
+// that BaseFact is a separate class from Fact because the RHS of conditional 
+// assertions can only contain BaseFacts.
+// Base facts have three forms:
+//  - A predicate
+//  - An attribute
+//  - A canActAs expression
+// The particular choice of form is represented with a variant which is 
+// encapsulated.
+class BaseFact {
+ public:
+  explicit BaseFact(Predicate predicate) : value_(std::move(predicate)) {};
+  explicit BaseFact(Attribute attribute) : value_(std::move(attribute)) {};
+  explicit BaseFact(CanActAs canActAs) : value_(std::move(canActAs)) {};
+  const std::variant<Predicate, Attribute, CanActAs>& value() const {
+    return value_;
+  }
+
+ private:
+  std::variant<Predicate, Attribute, CanActAs> value_;
+};
+
+// CanSay and Fact are forward-declared because they are mutually recursive.
+class CanSay;
+class Fact;
+
+// CanSay corresponds to an expression of the form <principal> canSay Fact
+class CanSay {
+  public:
+    explicit CanSay(Principal principal, std::unique_ptr<Fact>& fact)
+      : principal_(principal), fact_(std::move(fact)) {}
+    const Principal& principal() const { return principal_; }
+    const Fact* fact() const { return fact_.get(); }
+  private:
+    Principal principal_;
+    std::unique_ptr<Fact> fact_;
+};
+
+// Fact corresponds to either a base fact or a an expression of the form 
+// <principal> canSay 
+class Fact {
+ public:
+  explicit Fact(BaseFact base_fact) : value_(std::move(base_fact)) {}
+  explicit Fact(const CanSay* can_say)
```

Comment:
Is this `CanSay` not owned by the `Fact`? Could it be made into a `std::unique_ptr<CanSay>`, or is it pointed to elsewhere?

Reply:
fixed.